language/move-binary-format/src/serializer.rs

```diff
@@ -394,7 +394,22 @@ fn serialize_struct_handle(binary: &mut BinaryData, struct_handle: &StructHandle
     serialize_module_handle_index(binary, &struct_handle.module)?;
     serialize_identifier_index(binary, &struct_handle.name)?;
     serialize_ability_set(binary, struct_handle.abilities)?;
-    serialize_ability_sets(binary, &struct_handle.type_parameters)
+    serialize_ability_sets(binary, &struct_handle.type_parameters)?;
+    serialize_type_param_kinds(binary, &struct_handle.type_param_kinds)
+}
+
+fn serialize_type_param_kinds(binary: &mut BinaryData, kinds: &[TypeParamKind]) -> Result<()> {
+    for kind in kinds {
+        serialize_type_param_kind(binary, *kind)?;
+    }
+    Ok(())
+}
+
+fn serialize_type_param_kind(binary: &mut BinaryData, kind: TypeParamKind) -> Result<()> {
+    match kind {
+        TypeParamKind::Phantom => binary.push(SerializedTypeParamKindFlag::PHANTOM as u8),
+        TypeParamKind::NonPhantom => binary.push(SerializedTypeParamKindFlag::NON_PHANTOM as u8),
+    }
```

Comment:
Similarly, I think a discussion about [AoS (Array of Structs) vs SoA (Struct of Arrays)](https://en.wikipedia.org/wiki/AoS_and_SoA) should happen here. 

Do we want
```
[length] [abi 1] [abi 2] [abi 3] [phantom 1] [phantom 2] [phantom 3]
```
or
```
[length] [abi 1] [phantom 1] [abi 2] [phantom 2] [abi 3] [phantom 3]
```
?

Additionally 
- In the former case, do we want to compress the phantom bits into a bit vector?
- In the latter case, do we want to compress the phantom bit and the ability set into the same bit vector?

Reply:
The problem with "[length] [abi 1] [phantom 1] [abi 2] [phantom 2] [abi 3] [phantom 3]" is we don't want to make this a breaking change. 

So that gives you:
- If it is not present, it is assumed that all the type parameters are non-phantom
- if it is present, you specify the status/usage for each one

This way you don't need to make a breaking change... but then again you are going to have to gate this based on the version of the file format, so maybe the breaking change isn't a big deal and doing the AoS solution would work

Reply:
Yeah the mroe I think about it since you have to gate based on the file format version, we might as well just do the cleaner thing and do `[length] [abi 1] [phantom 1] [abi 2] [phantom 2] [abi 3] [phantom 3]`

Reply:
Yeah I originally also wanted compare the schemas in terms of binary compatibility, but then came into the same conclusion that any addition to the file format requires a new version number, making it meaningless to talk about backward compatibility, unless we come up with some notation that would allow us to say something like "version 5 is compatible with (can be read as) version 6" or equivalently introduce some sort of ".1" semantics, which I don't think we'll do. 

Thinking about binary compatibility might help make it easier for us to implement ser/de though, as minimizing the difference between the old and new format is one of the goals. Still, I don't think this is a major concern and I'd prefer that we simply take the new version as a chance to evolve the file format.